Optimization of HMM fitting.

lekcyjna123 opened the pull request:
In https://github.com/jmschrei/pomegranate/issues/802 i proposed adding transcoding of input sequences in hmm.fit to avoid invoking _check_input in hmm.summarize. This fix speed up fitting HMM models about two times.

With fix:
```
[1] Improvement: 253389.81343831721	Time (s): 0.2847
[2] Improvement: 213.1332476385287	Time (s): 0.2908
[3] Improvement: 76.93109403923154	Time (s): 0.2974
[4] Improvement: 44.39238843484782	Time (s): 0.283
[5] Improvement: 28.537198789825197	Time (s): 0.2837
[6] Improvement: 17.950176118465606	Time (s): 0.2804
[7] Improvement: 11.243089165887795	Time (s): 0.2813
[8] Improvement: 7.315842745301779	Time (s): 0.2797
Total Training Improvement: 253789.3164752493
Total Training Time (s): 2.9646
```

With pomegranate 0.13.4:
```
[1] Improvement: 253389.81343831721	Time (s): 0.71
[2] Improvement: 213.1332476385287	Time (s): 0.722
[3] Improvement: 76.93109403923154	Time (s): 0.7147
[4] Improvement: 44.39238843484782	Time (s): 0.7168
[5] Improvement: 28.537198789825197	Time (s): 0.7143
[6] Improvement: 17.950176118465606	Time (s): 0.7111
[7] Improvement: 11.243089165887795	Time (s): 0.7141
[8] Improvement: 7.315842745301779	Time (s): 0.7134
Total Training Improvement: 253789.3164752493
Total Training Time (s): 6.4269
```

lekcyjna123 commented:
I see, that not all test passed. I know already where is bug. I turned off checking input for al types of input in hmm.fit, but i don't make anything when i get DataGenerator as input data. So i hope that i will soon fix this merge request.

lekcyjna123 commented:
I fixed my request to pass all test. I would apologize for commits which didn't passed test. I have problem with testing this files locally. Nose return me that there are errors in release 0.13.4 and this error are shown in all my version, while test with errors in GitHub passed locally.

Now i have seen that accidentally i added to this request commit which fixed benchmark for hmm and i don't how i should split it from this request to new.

jmschrei commented:
Nice. That's a good find. However, I don't feel comfortable simply turning that option off. Maybe we should simply make `check_input` an argument the user can pass in and have appropriate documentation telling them to turn it off for speed?

lekcyjna123 commented:
Ok, i can try do this. 
>That way the user can turn it off when they know they have clean inputs.
We have to run `check_input` at least one time, because in this function we transcode arbitrary symbols given by user to `float64`.
>I was trying to support arbitrary inputs, including each batch being a new set of data that the mode hadn't seen before
Yes, this need running `check_input` each time, but i don't see where there is an option to pass into `fit` function data, which will be different in each step of fitting.

lekcyjna123 commented:
I implemented new argument `multiple_check_input` for `hmm.fit()` and `hmm.from_sample()` it has default value `True`. In default configuration this functions should be the same as in pomegranate 0.13.4. When we set this argument to `False` we `check_input` only once on the beggining of `hmm.fit()`. I added also 3 new test to test_hmm.py to test situation in which `multiple_check_input=False`. This test are simple copy of basic tests of fitting models extended with this new argument.

jmschrei commented:
Awesome. This is a great PR. Thanks!
